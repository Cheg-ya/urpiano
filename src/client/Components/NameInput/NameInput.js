import React, { Component } from 'react';
import { IoIosWarning } from 'react-icons/io';
import PropTypes from 'prop-types';
import './NameInput.scss';

class NameInput extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: '',
      invalid: false
    };

    this.onClickClose = this.onClickClose.bind(this);
    this.handleOnSubmit = this.handleOnSubmit.bind(this);
    this.handleOnChange = this.handleOnChange.bind(this);
  }

  onClickClose() {
    this.props.closeModal();
  }

  handleOnSubmit(e) {
    e.preventDefault();

    const { name, invalid } = this.state;

    if (name.length < 3 && !invalid) {
      return this.setState(() => {
        return {
          invalid: true
        };
      });
    }

    this.props.onCreate(name);
  }

  handleOnChange(e) {
    const name = e.target.value;

    this.setState(() => {
      return {
        name
      };
    });
  }

  render() {
    const { name, invalid } = this.state;

    return (
      <form className="nameInputForm" onSubmit={this.handleOnSubmit}>
        <div className="formLogo">UP</div>
        <i className="closeBtn fas fa-times-circle" onClick={this.onClickClose}></i>
        <div className="inputCover">
          <input
            type="text"
            className="nameInputField"
            placeholder="Type your name"
            value={name}
            onChange={this.handleOnChange}
          />
          {invalid && <label><IoIosWarning />At least 3 letters required</label>}
          <button className="submitBtn" type="submit">Submit</button>
        </div>
      </form>
    );
  }
}

export default NameInput;

NameInput.propTypes = {
  closeModal: PropTypes.func.isRequired,
  onCreate: PropTypes.func.isRequired
};
