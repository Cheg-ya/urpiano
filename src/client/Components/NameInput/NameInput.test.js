import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import NameInput from './NameInput';

describe('NameInput component test', () => {
  const closeModal = jest.fn().mockImplementation(() => 'close');
  const onCreate = jest.fn().mockImplementation(() => 'create');
  const initialState = {
    name: '',
    invalid: false
  };

  describe('render and props check', () => {
    const component = shallow((
      <NameInput closeModal={closeModal} onCreate={onCreate} />
    ));

    it('renders correctly with state', () => {
      const tree = renderer.create((
        <NameInput closeModal={closeModal} onCreate={onCreate} />
      )).toJSON();

      expect(tree).toMatchSnapshot();
      expect(component.exists()).toBe(true);
      expect(component.state()).not.toBeNull();
      expect(component.state()).toEqual(initialState);
      expect(component.state('name')).toBe('');
      expect(component.state('invalid')).toBe(false);
    });

    it('checks props type', () => {
      const component = mount(<NameInput closeModal={closeModal} onCreate={onCreate} />);
      const props = component.props();

      expect(typeof props.closeModal).toBe('function');
      expect(typeof props.onCreate).toBe('function');
    });
  });

  describe('custom methods test', () => {
    const onClickClose = jest.spyOn(NameInput.prototype, 'onClickClose');
    const handleOnChange = jest.spyOn(NameInput.prototype, 'handleOnChange');
    const handleOnSubmit = jest.spyOn(NameInput.prototype, 'handleOnSubmit');
    const component = mount(<NameInput closeModal={closeModal} onCreate={onCreate} />);
    const props = component.props();
    const tree = renderer.create((
      <NameInput closeModal={closeModal} onCreate={onCreate} />
    )).toJSON();

    it('calls onClickClose when click event occurred', () => {
      const closeModal = jest.spyOn(props, 'closeModal');
      component.find('.closeBtn').simulate('click');

      expect(onClickClose).toHaveBeenCalled();
      expect(onClickClose).toHaveBeenCalledTimes(1);
      expect(closeModal).toHaveBeenCalled();
      expect(closeModal).toHaveBeenCalledTimes(1);
      expect(tree).toMatchSnapshot();
    });

    it('calls handleOnChange when change event occurred', () => {
      component.find('.nameInputField').simulate('change', { target: { value: 'string' }});

      expect(handleOnChange).toHaveBeenCalled();
      expect(handleOnChange).toHaveBeenCalledTimes(1);
      expect(component.state('name')).toHaveLength(6);
      expect(tree).toMatchSnapshot();

      component.setState(initialState);
    });

    it('calls handleOnSubmit and changes visibility of label tag when submit event occurred', () => {
      const onCreate = jest.spyOn(props, 'onCreate');

      component.find('.nameInputForm').simulate('submit');

      expect(component.state('name')).toHaveLength(0);
      expect(handleOnSubmit).toHaveBeenCalled();
      expect(handleOnSubmit).toHaveBeenCalledTimes(1);
      expect(component.state('invalid')).toBe(true);
      expect(onCreate).not.toHaveBeenCalled();
      expect(component.find('.inputCover').children()).toHaveLength(3);
      expect(tree).toMatchSnapshot();

      component.setState({ name: 'string', invalid: false });
      component.find('.nameInputForm').simulate('submit');

      expect(handleOnSubmit).toHaveBeenCalled();
      expect(onCreate).toHaveBeenCalled();
      expect(onCreate).toHaveBeenCalledTimes(1);
      expect(component.find('.inputCover').children()).toHaveLength(2);
      expect(tree).toMatchSnapshot();

      component.unmount();
    });
  });
});
