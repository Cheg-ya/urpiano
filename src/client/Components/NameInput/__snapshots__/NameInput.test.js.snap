// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameInput component test custom methods test calls handleOnChange when change event occurred 1`] = `
<form
  className="nameInputForm"
  onSubmit={[Function]}
>
  <div
    className="formLogo"
  >
    UP
  </div>
  <i
    className="closeBtn fas fa-times-circle"
    onClick={[Function]}
  />
  <div
    className="inputCover"
  >
    <input
      className="nameInputField"
      onChange={[Function]}
      placeholder="Type your name"
      type="text"
      value=""
    />
    <button
      className="submitBtn"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`NameInput component test custom methods test calls handleOnSubmit and changes visibility of label tag when submit event occurred 1`] = `
<form
  className="nameInputForm"
  onSubmit={[Function]}
>
  <div
    className="formLogo"
  >
    UP
  </div>
  <i
    className="closeBtn fas fa-times-circle"
    onClick={[Function]}
  />
  <div
    className="inputCover"
  >
    <input
      className="nameInputField"
      onChange={[Function]}
      placeholder="Type your name"
      type="text"
      value=""
    />
    <button
      className="submitBtn"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`NameInput component test custom methods test calls handleOnSubmit and changes visibility of label tag when submit event occurred 2`] = `
<form
  className="nameInputForm"
  onSubmit={[Function]}
>
  <div
    className="formLogo"
  >
    UP
  </div>
  <i
    className="closeBtn fas fa-times-circle"
    onClick={[Function]}
  />
  <div
    className="inputCover"
  >
    <input
      className="nameInputField"
      onChange={[Function]}
      placeholder="Type your name"
      type="text"
      value=""
    />
    <button
      className="submitBtn"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`NameInput component test custom methods test calls onClickClose when click event occurred 1`] = `
<form
  className="nameInputForm"
  onSubmit={[Function]}
>
  <div
    className="formLogo"
  >
    UP
  </div>
  <i
    className="closeBtn fas fa-times-circle"
    onClick={[Function]}
  />
  <div
    className="inputCover"
  >
    <input
      className="nameInputField"
      onChange={[Function]}
      placeholder="Type your name"
      type="text"
      value=""
    />
    <button
      className="submitBtn"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`NameInput component test render and props check renders correctly with state 1`] = `
<form
  className="nameInputForm"
  onSubmit={[Function]}
>
  <div
    className="formLogo"
  >
    UP
  </div>
  <i
    className="closeBtn fas fa-times-circle"
    onClick={[Function]}
  />
  <div
    className="inputCover"
  >
    <input
      className="nameInputField"
      onChange={[Function]}
      placeholder="Type your name"
      type="text"
      value=""
    />
    <button
      className="submitBtn"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
